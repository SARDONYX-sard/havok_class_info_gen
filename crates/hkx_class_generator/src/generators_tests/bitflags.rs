use crate::{
    generators::one_class::bitflags::generate_bitflags,
    hkxcmd_parser::{hk_types::Type, Enum, EnumItem},
};
#[allow(unused)]
use pretty_assertions::assert_eq;

#[test]
fn should_generate_bitflags() {
    // Define an example enum info
    let enum_info = Enum {
        name: "FlagValues".into(),
        hk_type: Type::Enum,
        sub_type: Type::Int32,
        enum_item: vec![
            ("FLAGS_NONE", 0),
            ("ALIGN8", 1 << 7),
            ("ALIGN16", 1 << 8),
            ("NOT_OWNED", 1 << 9),
            ("SERIALIZE_IGNORED", 1 << 10),
        ]
        .into_iter()
        .map(|(tag, value)| EnumItem {
            name: tag.into(),
            value,
        })
        .collect(),
        flags: "00000000".into(),
    };
    let generated_code = generate_bitflags(&enum_info);

    insta::assert_snapshot!(generated_code);
}
