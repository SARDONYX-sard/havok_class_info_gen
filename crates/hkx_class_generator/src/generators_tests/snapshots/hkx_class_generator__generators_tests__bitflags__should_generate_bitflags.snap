---
source: crates/hkx_class_generator/src/generators_tests/bitflags.rs
expression: generated_code
---
bitflags::bitflags! {
    /// # Bit flags that represented enum.
    ///
    /// # Note
    /// The 0 flag is always enabled when the trailing bit is 0.
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
    pub struct FlagValues: u16 {
        /// Special flags." 0" is used when the character "0" is received.
        /// - only "0" is also entered during serialization.
        /// - This flag is the default value that exists for all flags.
        const NULL= !0;
        const FLAGS_NONE = 0;
        const ALIGN8 = 128;
        const ALIGN16 = 256;
        const NOT_OWNED = 512;
        const SERIALIZE_IGNORED = 1024;
    }
}

impl Default for FlagValues {
    fn default() -> Self {
        Self::NULL
    }
}

impl TryFrom<usize> for FlagValues {
    type Error = String;

    fn try_from(value: usize) -> Result<Self, Self::Error> {
        Self::from_bits(value as u16).ok_or_else(|| format!("Set invalid value: {value}"))
    }
}

impl TryFrom<u16> for FlagValues {
    type Error = String;

    fn try_from(value: u16) -> Result<Self, Self::Error> {
        Self::from_bits(value).ok_or_else(|| format!("Set invalid value: {value}"))
    }
}

impl serde::Serialize for FlagValues {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        if self.contains(Self::NULL) {
            serializer.serialize_str("0")
        } else {
            serializer.serialize_str(&self.human_readable())
        }
    }
}

impl<'de> serde::Deserialize<'de> for FlagValues {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let value = Option::<std::borrow::Cow<'de, str>>::deserialize(deserializer)?;

        match value {
            Some(s) => {
                if s.as_ref() == "0" {
                    return Ok(Self::NULL);
                };
                let mut flags = Self::empty();
                for token in s.split('|') {
                    match token.trim() {
                        "FLAGS_NONE" => flags |= Self::FLAGS_NONE,
                        "ALIGN8" => flags |= Self::ALIGN8,
                        "ALIGN16" => flags |= Self::ALIGN16,
                        "NOT_OWNED" => flags |= Self::NOT_OWNED,
                        "SERIALIZE_IGNORED" => flags |= Self::SERIALIZE_IGNORED,
                        unknown => match parse_int::parse(unknown) {
                            Ok(int) => {
                                if let Some(bits) = Self::from_bits(int) {
                                    flags |= bits
                                } else {
                                    return Err(serde::de::Error::custom(format!(
                                        "Expected FlagValues flags but got '{unknown}'",
                                    )));
                                };
                            }
                            Err(_err) => {
                                return Err(serde::de::Error::custom(format!(
                                    "Expected FlagValues flags or integer, but got '{unknown}'"
                                )))
                            }
                        },
                    }
                }
                Ok(flags)
            }
            None => Ok(Self::NULL),
        }
    }
}

impl core::fmt::Display for FlagValues {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "{}", self.bits())
    }
}

impl FlagValues {
    /// Use a string format that is easy for humans to read.
    ///
    /// Like this.
    /// `"FLAGS_NONE | SERIALIZE_IGNORED"`
    pub fn human_readable(&self) -> std::borrow::Cow<'_, str> {
        let mut flags = Vec::new();

        if self.contains(Self::NULL) {
            return "0".into();
        };

        if self.contains(Self::FLAGS_NONE) {
            flags.push("FLAGS_NONE");
        }
        if self.contains(Self::ALIGN8) {
            flags.push("ALIGN8");
        }
        if self.contains(Self::ALIGN16) {
            flags.push("ALIGN16");
        }
        if self.contains(Self::NOT_OWNED) {
            flags.push("NOT_OWNED");
        }
        if self.contains(Self::SERIALIZE_IGNORED) {
            flags.push("SERIALIZE_IGNORED");
        }

        flags.join("|").into()
    }
}
